driver/base/response.go

```diff
@@ -95,17 +98,50 @@ func (r *Response) Record(rawResult []byte, result string) {
 }
 
 // TextFsmParse parses recorded output w/ a provided textfsm template.
-func (r *Response) TextFsmParse(template string) ([]map[string]interface{}, error) {
-	t, err := os.ReadFile(template)
-	if err != nil {
-		logging.LogError(
-			r.FormatLogMessage(
-				"error",
-				fmt.Sprintf("Failed opening provided template, error: %s\n", err.Error()),
-			),
-		)
+// the argument is interpreted as URL or filesystem path, for example:
+// response.TextFsmParse("http://example.com/textfsm.template") or
+// response.TextFsmParse("./local/textfsm.template")
+func (r *Response) TextFsmParse(path string) ([]map[string]interface{}, error) {
+	var t []byte
+	var err error
+	switch {
+	case strings.HasPrefix(path, "http://") || strings.HasPrefix(path, "https://"):
+		resp, err := http.Get(path)
+		if err != nil {
+			logging.LogError(
+				r.FormatLogMessage(
+					"error",
+					fmt.Sprintf("Failed downloading template, error: %s\n", err.Error()),
+				),
+			)
+			return []map[string]interface{}{}, ErrFailedOpeningTemplate
+		}
+		defer resp.Body.Close()
+
+		t, err = io.ReadAll(resp.Body)
+		if err != nil {
+			logging.LogError(
+				r.FormatLogMessage(
+					"error",
+					fmt.Sprintf("Failed reading downloaded template, error: %s\n", err.Error()),
+				),
+			)
+			return []map[string]interface{}{}, ErrFailedOpeningTemplate
```

Comment:
I wonder @carlmontanari why dont you return nil in the error case?

Reply:
thanks for the review man. what would be the case of returning `nil` as a value here?

Reply:
What I was getting to is typically when there is an error, you don't return an empty pointer type value, but a nil value

https://www.reddit.com/r/golang/comments/6ffls4/return_nil_or_pointer_to_empty_struct_on_error/

Reply:
imho, if there's an error,  you're supposed to check it.
and empty struct saves you from nil pointer exception if you forget to check it.

Reply:
> if there's an error, you're supposed to check it.

check what?

I am talking about line 129, where @carlmontanari currently has
```go
return []map[string]interface{}{}, ErrFailedOpeningTemplate
```

in 99% cases the caller checks the error value, and not the empty slice of maps. It would make sense to return an empty slice only in the case when it is designed to hold some value even when the error is not nil.
But this is not our case, so I do not see a reason to not return `nil` as a first value

Reply:
> check what?
check the returned error. this is Go's idiom. 
> in 99% cases the caller checks the error value, and not the empty slice of maps
exactly right
> It would make sense to return an empty slice only in the case when it is designed to hold some value
I don't agree with this. My thinking is: you check the err first as per the standard. If err == nil, then it's safe to access whatever was returned as the first argument. 
If the first arg is a `nil` then you have to take special care about it, which defeats the purpose of having an error as the 2nd argument.

Reply:
> My thinking is: you check the err first as per the standard. If err == nil, then it's safe to access whatever was returned as the first argument.

yes, exactly. I am saying that if you have an error (err != nil), then there is no sense in returning an empty slice, it should be nil, because you don't inspect the non error value.
both switch cases return some error (lines 129 and 142), so the caller will always have a non nil error. The caller will treat the error, but he doesn't need an empty slice as the first arg, it should be nil imho

Reply:
yep, it makes sense. but I still don't get the benefit of returning nil over an empty slice in this case. i may be missing something but to me it doesn't make any difference if i get a nil or an empty slice together with my non-nil error.

Reply:
yes, it is purely idiomatic vs non-idiomatic. there is no perf benefit in doing one way over the other.

I wanted to quickly find refs that will speak to my point of returning `nil` over an empty slice/map/interface, and the first one was that reddit post 

https://www.reddit.com/r/golang/comments/6ffls4/return_nil_or_pointer_to_empty_struct_on_error/

a few others (apart from existing code bases)

* https://morioh.com/p/385f6171e967
* https://www.reddit.com/r/golang/comments/4asbwo/difference_between_return_nil_err_and_return_t_err/

Reply:
Dang, ya'll been busy! :) In this case there is a very simple answer (love it when that happens). In scrapli (py) and because of that, in scrapligo -- when parsing with textfsm (or genie or ttp in py version), if the parsing fails for whatever reason, we *always return a list/dict*. Especially in python-land the other option was basically to return the original payload that got sent to the parser. But that is super stupid/vague IMO because you can then try to do like `.items()` or something and obviously that will fail on a list. 

So.... while I generally care about being idiomatic - if for no other reason than to use conversations like this to learn and try to write better code - in this case I think id rather stick w/ returning the empty map to keep parity with python version.

Thoughts?

Reply:
now that I've caught up on ya'lls novel ill actually look at the pr 😁

Reply:
Also, thanks a ton for giving us some extra eyes @hellt  --- always appreciated!

Reply:
> So.... while I generally care about being idiomatic - if for no other reason than to use conversations like this to learn and try to write better code - in this case I think id rather stick w/ returning the empty map to keep parity with python version.

So writing Python in Go, okay :)

Reply:
hahah. breakin my balls as per usual!! I know that you are *right* but... parity. I suppose it isn't a big deal either way.

like we said in chat -- don't worry, boxen will be more idiomatic always with no python stink creeping in ;)